src/Polysemy/Internal.hs

```diff
@@ -267,7 +267,7 @@ instance Functor (Sem f) where
 
 
 instance Applicative (Sem f) where
-  pure a = Sem $ const $ pure a
+  pure a = Sem $ \_ -> pure a
```

Comment:
?

Reply:
Didn't mean for this comment to be submitted. There _are_ some minor changes here and there I'm not sure why you've made, but they shouldn't affect anything.

Reply:
All of those changes were prompted by GHC complaining. In this case, something about polytypes

Reply:
```
    • Couldn't match type ‘b0’
                     with ‘forall x. Union f (Sem f) x -> m x’
      Expected: (forall x. Union f (Sem f) x -> m x) -> m a
        Actual: b0 -> m a
      Cannot instantiate unification variable ‘b0’
      with a type involving polytypes: forall x. Union f (Sem f) x -> m x
```

Reply:
Oh! This is a side-effect of quick look impredicativity (or the loss of function contravariance?), I wager. In earlier versions, I bet `b0` was instantiated to `Union f (Sem f) x0 -> m x0` for some ambiguous `x0`, making it first-rank and thus valid. Now, `b0` is instead instantiated to the proper impredicative `(forall x. Union f (Sem f) x -> m x)`, which is rejected since quick look impredicativity isn't on.

Reply:
should we activate it conditionally?

Reply:
FYI works with `ImpredicativeTypes`

Reply:
Are you running with GHC 9.2? Otherwise, quick-look isn't in yet.

If you are running with 9.2, why not. I expect `ImpredicativeTypes` to replace `RankNTypes` entirely.

Reply:
no, 9.0.1. Still fixes the error!

Reply:
Ergh, actually, we want backwards compatibility, so I'd wait until GHC 9.6 before enabling it (since by then the "three most recent major versions" standard will not include any version prior to 9.2).

Avoid `ImpredicativeTypes` for the moment, unless you're willing to guard its use with CPP.

Reply:
that's what I meant with "conditionally". I'm also using CPP to compensate for the breaking change in `TH.Syntax`, so I would find it acceptable to be able to use the same code as before by using the extension

Reply:
`ImpredicativeTypes` in 9.0.1 is still undefined. Don't use it before 9.2.

Reply:
ok, so we keep this change

Reply:
Yup.